{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fcb65477",
   "metadata": {},
   "source": [
    "# Different variable types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71562fb7",
   "metadata": {},
   "source": [
    "Double quote and single quote are the same. Sometimes you may want to use different quote in a string to avoid confusion (fot example, you want to show a string in a string)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d96d8396",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Strings (for sequences, gene names, etc.)\n",
    "gene_name = \"TP53\" \n",
    "dna_sequence = 'ATGCGATACGCTTGA' \n",
    "print(f\"Gene: {gene_name}, Sequence start: {dna_sequence[:10]}\") # the first 10 characters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec29eabf",
   "metadata": {},
   "source": [
    "`f\"xxx{} xxx\"` is a string formatting method. It is the same as `\"xxx\" + str(xxx) + \"xxx\"`\n",
    "\n",
    "or `\"xxx%s xxx\" % xxx`\n",
    "'s' is the variable, no brackets needed\n",
    "or  `\"xxx{0} xxx\".format(xxx)`\n",
    "f is replaced by .format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ff0b83c-2306-4e5c-91f4-8112424fc734",
   "metadata": {},
   "outputs": [],
   "source": [
    "!!!My code!!!\n",
    "#String of data - single piece of data, can be listed but will only consider one etry at a time; can use both double / single quotes\n",
    "    gene = \"gene_name1\", \"gene_name2\", \"gene_name3\"\n",
    "    dna_sequence = \"ATCGCGC\", \"AGGCCTA\", \"ATTCGGG\"\n",
    "\n",
    "#printing text with you variable in the text\n",
    "print(f\"Many genes are expressed, for example {gene} which starts with sequence {dna_sequence}\") \n",
    "\n",
    "#if want to list the first 2 instead of all 3\n",
    "print(f\"Many genes are expressed, for example {gene[:2]} which start with sequence {dna_sequence[:2]}\") #gives only first 2 entries\n",
    "\n",
    "Three string formats to use:\n",
    "- f-string\n",
    "- str()\n",
    "- %s\n",
    "- .format()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3545368",
   "metadata": {},
   "outputs": [],
   "source": [
    "Exercise - changing different formats of string:\n",
    "print(f\"Many genes are expressed, for example {gene} which starts with sequence {dna_sequence}\") \n",
    "print(\"Many genes are expressed, for example \" + str(gene) + \" which starts with sequence \" + str(dna_sequence)) <- change this into str() format\n",
    "print(\"Many genes are expressed, for example %s which starts with sequence %s\" % (gene, dna_sequence)) <- change this into % form\n",
    "print(\"Many genes are expressed, for example {gene_name} which starts with sequence {seq}\".format(gene_name = gene, seq = dna_sequence)) <- change this into .format form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11280a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Integers (for positions, counts, etc.) numbers as categorical variables\n",
    "chromosome_number = 17\n",
    "exon_count = 11\n",
    "print(f\"Chromosome {chromosome_number} has {exon_count} exons\")\n",
    "\n",
    "# Floats (for measurements, concentrations, etc.) continuous variables\n",
    "ph_value = 7.4\n",
    "concentration = 0.05  # 50 mM\n",
    "print(f\"pH: {ph_value}, Concentration: {concentration}M\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fcdb125",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Booleans (for conditions) binary; 0 = False, 1 = True\n",
    "is_expressed = True\n",
    "is_mutated = False\n",
    "print(f\"Gene expressed: {is_expressed}, Mutated: {is_mutated}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9be9d0b-d1b1-4604-ba7c-3775b7fe8483",
   "metadata": {},
   "outputs": [],
   "source": [
    "!!!My code!!!\n",
    "Can put all three types of data: character data, integers, floats, booleans into one print sentence using f-string format\n",
    "\n",
    "gene = \"gene_name1\", \"gene_name2\", \"gene_name3\" #character\n",
    "chromosome = 2, 13, 22      #integer\n",
    "expression_level = 0.45, 0.85, 0.024 #float\n",
    "mutated = True, False, True\n",
    "\n",
    "print(f\"Many genes are expressed, for example {gene} from chromosome {chromosome} at expression level {expression_level}, this gene is mutated {mutated}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae4452d8",
   "metadata": {},
   "source": [
    "# list v.s tuple v.s set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bfe22da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lists (for collections of items)\n",
    "genes = [\"BRCA1\", \"BRCA2\", \"TP53\", \"EGFR\"]\n",
    "print(f\"Gene list: {genes}\")\n",
    "\n",
    "# Tuples (immutable sequences)\n",
    "chromosome = (17, \"p13.1\")  # (chromosome number, cytogenetic band)\n",
    "print(f\"Chromosome location: {chromosome}\")\n",
    "\n",
    "# Sets (unique collections)\n",
    "unique_bases = set(dna_sequence)\n",
    "print(f\"Unique bases in sequence: {unique_bases}\")\n",
    "\n",
    "#Difference between tuples and sets: cant change elements in tuples later but still can in sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b241ef58-cddf-4322-ad6b-56aa3626e6e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "!!!My code!!!\n",
    "#Differencee between strings vs lists vs tuples vs sets\n",
    "\n",
    "String: gene = \"gene_name1\", \"gene_name2\", \"gene_name3\"\n",
    "List: genes = [\"gene_name1\", \"gene_name2\", \"gene_name3\"]\n",
    "Tuple: genes_expression = (\"gene_name1\", 0.45, \"gene_name2\", 0.86, \"gene_name3\", 0.024)\n",
    "Set: unique_seq = set(dna_sequence) #creates a set of unique caharcters from a string\n",
    "#if dna_sequence = \"A\", \"C\", \"C\", \"C\" -> set(dna_sequence) makes it into \"A\", \"C\", \"C\", \"C\" => \"ACCC\"\n",
    "\n",
    "print (f\"Many genes are expressed, for example {gene}\")\n",
    "print (f\"Many genes are expressed, for example {genes}\")\n",
    "print (f\"Many genes are expressed at different levels, for example {genes_expression})\n",
    "print (f\"Many genes are expressed and some have unique sequences such as {unique_seq})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfa4aac5",
   "metadata": {},
   "source": [
    "<font size=4 color='red'>python is 0 indexed !!!</font> So, the first element in a list is at index ***0***."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea6887fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of genes in the list 4 the first gene is  BRCA1\n"
     ]
    }
   ],
   "source": [
    "# to index a value in a list\n",
    "print(\"the number of genes in the list :\", len(genes), \n",
    "      \"the first gene is \", genes[0]) # the first value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5309f604",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene: TP53, Sequence start: ATGCGATACG\n",
      "Chromosome 17 has 11 exons\n",
      "pH: 7.4, Concentration: 0.05M\n",
      "Gene expressed: True, Mutated: False\n",
      "Gene list: ['BRCA1', 'BRCA2', 'TP53', 'EGFR']\n",
      "Chromosome location: (17, 'p13.1')\n",
      "Gene info: {'name': 'TP53', 'chromosome': 17, 'function': 'tumor suppressor', 'length': 19149}\n",
      "Unique bases in sequence: {'G', 'C', 'A', 'T'}\n"
     ]
    }
   ],
   "source": [
    "# Dictionaries (key-value pairs, like hash tables)\n",
    "gene_info = {\n",
    "    \"name\": \"TP53\",\n",
    "    \"chromosome\": 17,\n",
    "    \"function\": \"tumor suppressor\",\n",
    "    \"length\": 19149\n",
    "}\n",
    "print(f\"Gene info: {gene_info}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5f36dba-705c-40ec-afd0-70234a2a5aa0",
   "metadata": {},
   "source": [
    "!!!My code!!!\n",
    "Create a gene_info table\n",
    "gene1_info = {\n",
    "\"name\": \"OTX2\",\n",
    "\"chromosome\": 14,\n",
    "\"function\": \"embryonic development\",\n",
    "\"length\": 16789\n",
    "}\n",
    "print (f\"OTX2 gene info: {gene_info1}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae7c602e",
   "metadata": {},
   "source": [
    "# if-else statements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d9739d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic if-else\n",
    "expression_level = 1500\n",
    "threshold = 1000\n",
    "\n",
    "if expression_level > threshold:\n",
    "    print(\"Gene is highly expressed\")\n",
    "elif expression_level > 500:\n",
    "    print(\"Gene has moderate expression\")\n",
    "else:\n",
    "    print(\"Gene has low expression\")\n",
    "\n",
    "# Multiple conditions (important for bioinformatics filters)\n",
    "read_depth = 30\n",
    "quality_score = 20\n",
    "is_variant = True\n",
    "\n",
    "if read_depth >= 20 and quality_score >= 15 and is_variant:\n",
    "    print(\"Variant passes quality filters\")\n",
    "else:\n",
    "    print(\"Variant fails quality filters\")\n",
    "\n",
    "# Using 'in' for sequence checking (very useful in bioinformatics!)\n",
    "codon = \"ATG\"\n",
    "\n",
    "if codon in [\"ATG\", \"GTG\", \"TTG\"]:\n",
    "    print(f\"{codon} is a start codon\")\n",
    "elif codon in [\"TAA\", \"TAG\", \"TGA\"]:\n",
    "    print(f\"{codon} is a stop codon\")\n",
    "else:\n",
    "    print(f\"{codon} is a regular codon\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66048a94",
   "metadata": {},
   "source": [
    "# try-except "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3e02d752",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (974286641.py, line 7)",
     "output_type": "error",
     "traceback": [
      "  \u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[1]\u001b[39m\u001b[32m, line 7\u001b[39m\n\u001b[31m    \u001b[39m\u001b[31mprint(\"hello world\"+ !)\u001b[39m\n                         ^\n\u001b[31mSyntaxError\u001b[39m\u001b[31m:\u001b[39m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    print(hello_world)\n",
    "except NameError:\n",
    "    print(\"invalid variable name\")\n",
    "\n",
    "try:\n",
    "    print(\"hello world\"+ !)\n",
    "except NameError:\n",
    "    print(\"invalid variable name\")\n",
    "\n",
    "\n",
    "try:\n",
    "    print(\"hello world\"+ !)\n",
    "except:\n",
    "    print(\"invalid printing content\")\n",
    "\n",
    "#For handling errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "12362806-fbad-4104-8f8c-2dbe4f6f122d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "invalid variable name\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    print(hello_world)\n",
    "except NameError:\n",
    "    print(\"invalid variable name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4780971f-4f10-4e21-98aa-52d1e7c3f122",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (695013279.py, line 2)",
     "output_type": "error",
     "traceback": [
      "  \u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[3]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[31m    \u001b[39m\u001b[31mprint(\"hello world\"+ !)\u001b[39m\n                         ^\n\u001b[31mSyntaxError\u001b[39m\u001b[31m:\u001b[39m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    print(\"hello world\"+ !)\n",
    "except NameError:\n",
    "    print(\"invalid variable name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "02ac3eb6-ff81-4524-91b8-eb7efbb21190",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is a NameError\n",
      "This is a TypeError\n",
      "invalid operation\n",
      "invalid operation\n"
     ]
    }
   ],
   "source": [
    "!!!My code!!!\n",
    "#try-except function for catching errors, allowing the job not to crash if an error is encounteres - but only works for NameError / TypeError / ZeroDivisionError not SyntaxError\n",
    "\n",
    "#NameError\n",
    "try:\n",
    "    print(hello)\n",
    "except NameError:\n",
    "    print(\"This is a NameError\")\n",
    "\n",
    "#TypeError\n",
    "try:\n",
    "    print(\"hello world\" + 5)\n",
    "except TypeError:\n",
    "    print(\"This is a TypeError\")\n",
    "\n",
    "#TypeError + ZeroDivisionError\n",
    "try:\n",
    "    print(\"hello world\" / 0)\n",
    "except (TypeError, ZeroDivisionError):\n",
    "    print(\"invalid operation\")\n",
    "\n",
    "#prevention of all errors\n",
    "try:\n",
    "    print(hello / 0)\n",
    "except (NameError, TypeError, ZeroDivisionError):\n",
    "    print(\"invalid operation\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee676d3b",
   "metadata": {},
   "source": [
    "# Loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ba467069",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino acid positions:\n",
      "Position 1: M\n",
      "Position 2: V\n",
      "Position 3: L\n",
      "Position 4: S\n",
      "Position 5: P\n",
      "Position 6: A\n",
      "Position 7: D\n",
      "Position 8: K\n",
      "Position 9: T\n",
      "Position 10: N\n",
      "Position 11: V\n",
      "Position 12: K\n",
      "Position 13: A\n",
      "Position 14: A\n",
      "Position 15: W\n",
      "Position 16: G\n",
      "Position 17: K\n",
      "Position 18: V\n",
      "Position 19: G\n",
      "Position 20: A\n",
      "Position 21: H\n",
      "Position 22: A\n",
      "Position 23: G\n",
      "Position 24: E\n",
      "Position 25: Y\n",
      "Position 26: G\n",
      "Position 27: A\n",
      "Position 28: E\n",
      "Position 29: A\n",
      "Position 30: L\n",
      "Position 31: E\n",
      "Position 32: R\n",
      "Position 33: M\n",
      "Position 34: F\n",
      "Position 35: L\n",
      "Position 36: S\n",
      "Position 37: F\n",
      "Position 38: P\n",
      "Position 39: T\n",
      "Position 40: T\n",
      "Position 41: K\n",
      "Position 42: T\n",
      "Position 43: Y\n",
      "Position 44: F\n",
      "Position 45: P\n",
      "Position 46: H\n",
      "Position 47: F\n",
      "\n",
      "First 10 amino acids:\n",
      "1: M\n",
      "2: V\n",
      "3: L\n",
      "4: S\n",
      "5: P\n",
      "6: A\n",
      "7: D\n",
      "8: K\n",
      "9: T\n",
      "10: N\n",
      "\n",
      "Processing FASTA sequences:\n",
      "Processing: ATCG\n",
      "Processing: GCTA\n",
      "Processing: TTAA\n",
      "Processing: CCGG\n",
      "\n",
      "Finding first methionine (M) in protein:\n",
      "Found methionine at position 1\n",
      "\n",
      "DNA: ATGCGATACGCTTGA\n",
      "RNA: AUGCGAUACGCUUGA\n"
     ]
    }
   ],
   "source": [
    "# For loop - iterate through sequences\n",
    "protein_sequence = \"MVLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF\"\n",
    "\n",
    "print(\"Amino acid positions:\")\n",
    "for i, aa in enumerate(protein_sequence):\n",
    "    print(f\"Position {i+1}: {aa}\")\n",
    "\n",
    "# For loop with range\n",
    "print(\"\\nFirst 10 amino acids:\")\n",
    "for i in range(10):\n",
    "    print(f\"{i+1}: {protein_sequence[i]}\")\n",
    "\n",
    "# While loop - useful for parsing until conditions met\n",
    "fasta_sequences = [\"ATCG\", \"GCTA\", \"TTAA\", \"CCGG\", \"END\"]\n",
    "index = 0\n",
    "\n",
    "print(\"\\nProcessing FASTA sequences:\")\n",
    "while fasta_sequences[index] != \"END\":\n",
    "    print(f\"Processing: {fasta_sequences[index]}\")\n",
    "    index += 1\n",
    "\n",
    "# Loop control with break and continue\n",
    "print(\"\\nFinding first methionine (M) in protein:\")\n",
    "for i, aa in enumerate(protein_sequence):\n",
    "    if aa == \"M\":\n",
    "        print(f\"Found methionine at position {i+1}\")\n",
    "        break  # Exit loop early\n",
    "\n",
    "# List comprehension (Pythonic way to create lists)\n",
    "dna_sequence = \"ATGCGATACGCTTGA\"\n",
    "rna_sequence = [base.replace('T', 'U') for base in dna_sequence]\n",
    "print(f\"\\nDNA: {dna_sequence}\")\n",
    "print(f\"RNA: {''.join(rna_sequence)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7260c75b-998b-4c36-8409-c2dd34a7db99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n"
     ]
    }
   ],
   "source": [
    "#!!!My code!!!\n",
    "#Different loops: for / while\n",
    "\n",
    "#for loop (used when you know what you want to loop over)\n",
    "\n",
    "#ex.\n",
    "for i in range(5):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "922c9d82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino acid positions:\n",
      "Position 1: M\n",
      "Position 2: V\n",
      "Position 3: L\n",
      "Position 4: S\n",
      "Position 5: P\n",
      "Position 6: A\n",
      "Position 7: D\n",
      "Position 8: K\n",
      "Position 9: T\n",
      "Position 10: N\n",
      "Position 11: V\n",
      "Position 12: K\n",
      "Position 13: A\n",
      "Position 14: A\n",
      "Position 15: W\n",
      "Position 16: G\n",
      "Position 17: K\n",
      "Position 18: V\n",
      "Position 19: G\n",
      "Position 20: A\n",
      "Position 21: H\n",
      "Position 22: A\n",
      "Position 23: G\n",
      "Position 24: E\n",
      "Position 25: Y\n",
      "Position 26: G\n",
      "Position 27: A\n",
      "Position 28: E\n",
      "Position 29: A\n",
      "Position 30: L\n",
      "Position 31: E\n",
      "Position 32: R\n",
      "Position 33: M\n",
      "Position 34: F\n",
      "Position 35: L\n",
      "Position 36: S\n",
      "Position 37: F\n",
      "Position 38: P\n",
      "Position 39: T\n",
      "Position 40: T\n",
      "Position 41: K\n",
      "Position 42: T\n",
      "Position 43: Y\n",
      "Position 44: F\n",
      "Position 45: P\n",
      "Position 46: H\n",
      "Position 47: F\n"
     ]
    }
   ],
   "source": [
    "#ex. numbering aa of a sequence\n",
    "protein_sequence = \"MVLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF\"\n",
    "\n",
    "print(\"Amino acid positions:\")\n",
    "for i, aa in enumerate(protein_sequence):\n",
    "    print(f\"Position {i+1}: {aa}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1ac7242c-f2b6-4bc9-b6ce-a8471b89c31b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n"
     ]
    }
   ],
   "source": [
    "#while loop (used when you continue loop until a condition is met - ex output becomes false)\n",
    "\n",
    "#ex. adding +1 starting form 0 until it reaches 5\n",
    "count = 0\n",
    "\n",
    "while count < 5:\n",
    "    print(count)\n",
    "    count = count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "76a5990a-15ab-4671-a491-d9b5488000a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "START\n",
      "count is 0\n",
      "count is 1\n",
      "count is 2\n",
      "DONE\n"
     ]
    }
   ],
   "source": [
    "#ex. \n",
    "print(\"START\")\n",
    "count = 0\n",
    "while count < 3:\n",
    "    print(\"count is\", count)\n",
    "    count = count + 1\n",
    "print(\"DONE\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b2efafd",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "43cb163e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC content of 'ATGCGATACGCTTGA': 46.67%\n",
      "DNA: ATGCGATACGCTTGA\n",
      "RNA: AUGCGAUACGCUUGA\n",
      "RNA from reverse complement: UCAAGCGUAUCGCAU\n",
      "\n",
      "Sequence statistics: Length=15, GC=46.67%\n"
     ]
    }
   ],
   "source": [
    "# Basic function\n",
    "def calculate_gc_content(sequence):\n",
    "    \"\"\"Calculate GC content of a DNA sequence.\"\"\" #3 double quotes can write long multiline comments, common to use \"\"\" next to definition to tell what this function is about\n",
    "    sequence = sequence.upper()  # Convert to uppercase\n",
    "    gc_count = sequence.count('G') + sequence.count('C')\n",
    "    total_bases = len(sequence)\n",
    "    \n",
    "    if total_bases == 0: #== is judgement, = assigning value\n",
    "        return 0\n",
    "    \n",
    "    gc_percentage = (gc_count / total_bases) * 100\n",
    "    return gc_percentage\n",
    "\n",
    "# Test the function\n",
    "test_sequence = \"ATGCGATACGCTTGA\"\n",
    "gc_result = calculate_gc_content(test_sequence)\n",
    "print(f\"GC content of '{test_sequence}': {gc_result:.2f}%\")\n",
    "\n",
    "# Function with default parameters\n",
    "def transcribe_dna_to_rna(dna_sequence, reverse_complement=False):\n",
    "    \"\"\"Transcribe DNA to RNA, optionally from reverse complement.\"\"\"\n",
    "    if reverse_complement:\n",
    "        # Create reverse complement\n",
    "        complement = {'A': 'U', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "        rna = ''.join(complement.get(base, 'N') for base in reversed(dna_sequence))\n",
    "    else:\n",
    "        # Simple transcription (T -> U)\n",
    "        rna = dna_sequence.replace('T', 'U')\n",
    "    \n",
    "    return rna\n",
    "\n",
    "# Test transcription functions\n",
    "dna_seq = \"ATGCGATACGCTTGA\"\n",
    "print(f\"DNA: {dna_seq}\")\n",
    "print(f\"RNA: {transcribe_dna_to_rna(dna_seq)}\")\n",
    "print(f\"RNA from reverse complement: {transcribe_dna_to_rna(dna_seq, reverse_complement=True)}\")\n",
    "\n",
    "# Function returning multiple values\n",
    "def analyze_sequence(sequence):\n",
    "    \"\"\"Analyze a DNA sequence and return multiple statistics.\"\"\"\n",
    "    length = len(sequence)\n",
    "    gc_content = calculate_gc_content(sequence)\n",
    "    a_count = sequence.upper().count('A')\n",
    "    t_count = sequence.upper().count('T')\n",
    "    g_count = sequence.upper().count('G')\n",
    "    c_count = sequence.upper().count('C')\n",
    "    \n",
    "    return length, gc_content, a_count, t_count, g_count, c_count\n",
    "\n",
    "# Unpack multiple return values\n",
    "seq_stats = analyze_sequence(\"ATGCGATACGCTTGA\")\n",
    "print(f\"\\nSequence statistics: Length={seq_stats[0]}, GC={seq_stats[1]:.2f}%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9287eac8",
   "metadata": {},
   "source": [
    "# Class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7363aa8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene TP53 on chr17: 7668402-7687550\n",
      "Gene length: 27 bp\n",
      "GC content: 48.15%\n",
      "RNA transcript: AUGCGAUACGCUUGAAGCUA...\n",
      "Coding region starts with: ATGCGATACG\n",
      "\n",
      "Protein p53: 41 aa, ~4510 Da\n",
      "Lysine (K) positions: [8, 12, 17, 41]\n",
      "Contains 'SPAD' motif: True\n"
     ]
    }
   ],
   "source": [
    "# Basic class for a Gene\n",
    "class Gene:\n",
    "    \"\"\"A class to represent a gene.\"\"\"\n",
    "    \n",
    "    def __init__(self, name, sequence, chromosome, start_position, end_position):\n",
    "        \"\"\"Initialize a gene with its properties.\"\"\"\n",
    "        self.name = name\n",
    "        self.sequence = sequence.upper()  # Store as uppercase\n",
    "        self.chromosome = chromosome\n",
    "        self.start_position = start_position\n",
    "        self.end_position = end_position\n",
    "        self.length = len(sequence)\n",
    "    \n",
    "    def calculate_gc_content(self):\n",
    "        \"\"\"Calculate GC content of the gene.\"\"\"\n",
    "        gc_count = self.sequence.count('G') + self.sequence.count('C')\n",
    "        return (gc_count / self.length) * 100 if self.length > 0 else 0\n",
    "    \n",
    "    def transcribe(self):\n",
    "        \"\"\"Transcribe DNA to RNA.\"\"\"\n",
    "        return self.sequence.replace('T', 'U')\n",
    "    \n",
    "    def get_coding_region(self, start_codon=\"ATG\"):\n",
    "        \"\"\"Find coding region starting from start codon.\"\"\"\n",
    "        start_idx = self.sequence.find(start_codon)\n",
    "        if start_idx != -1:\n",
    "            # Return from start codon to end\n",
    "            return self.sequence[start_idx:]\n",
    "        return \"\"\n",
    "    \n",
    "    def __str__(self):\n",
    "        \"\"\"String representation of the gene.\"\"\"\n",
    "        return f\"Gene {self.name} on chr{self.chromosome}: {self.start_position}-{self.end_position}\"\n",
    "    \n",
    "    def __len__(self):\n",
    "        \"\"\"Return length of gene.\"\"\"\n",
    "        return self.length\n",
    "\n",
    "# Create a gene object\n",
    "tp53_gene = Gene(\n",
    "    name=\"TP53\",\n",
    "    sequence=\"ATGCGATACGCTTGAAGCTAGCTAGCT\",\n",
    "    chromosome=17,\n",
    "    start_position=7668402,\n",
    "    end_position=7687550\n",
    ")\n",
    "\n",
    "# Use the gene object\n",
    "print(tp53_gene)\n",
    "print(f\"Gene length: {len(tp53_gene)} bp\")\n",
    "print(f\"GC content: {tp53_gene.calculate_gc_content():.2f}%\")\n",
    "print(f\"RNA transcript: {tp53_gene.transcribe()[:20]}...\")\n",
    "print(f\"Coding region starts with: {tp53_gene.get_coding_region()[:10]}\")\n",
    "\n",
    "# More complex class for a Protein\n",
    "class Protein:\n",
    "    \"\"\"A class to represent a protein.\"\"\"\n",
    "    \n",
    "    def __init__(self, name, amino_acid_sequence):\n",
    "        self.name = name\n",
    "        self.sequence = amino_acid_sequence\n",
    "        self.length = len(amino_acid_sequence)\n",
    "        self.molecular_weight = self.calculate_molecular_weight()\n",
    "    \n",
    "    def calculate_molecular_weight(self):\n",
    "        \"\"\"Calculate approximate molecular weight in Daltons.\"\"\"\n",
    "        # Simple approximation: average amino acid weight = 110 Da\n",
    "        return self.length * 110\n",
    "    \n",
    "    def find_amino_acid_position(self, aa):\n",
    "        \"\"\"Find all positions of a specific amino acid.\"\"\"\n",
    "        positions = []\n",
    "        for i, residue in enumerate(self.sequence):\n",
    "            if residue == aa:\n",
    "                positions.append(i + 1)  # 1-based indexing\n",
    "        return positions\n",
    "    \n",
    "    def has_motif(self, motif):\n",
    "        \"\"\"Check if protein contains a specific motif.\"\"\"\n",
    "        return motif in self.sequence\n",
    "    \n",
    "    def __str__(self):\n",
    "        return f\"Protein {self.name}: {self.length} aa, ~{self.molecular_weight:.0f} Da\"\n",
    "\n",
    "# Create a protein object\n",
    "p53_protein = Protein(\"p53\", \"MVLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTK\")\n",
    "print(f\"\\n{p53_protein}\")\n",
    "print(f\"Lysine (K) positions: {p53_protein.find_amino_acid_position('K')}\")\n",
    "print(f\"Contains 'SPAD' motif: {p53_protein.has_motif('SPAD')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "828d30bc-caa0-4962-8473-ae98e84760f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello Alex\n"
     ]
    }
   ],
   "source": [
    "#!!! My code!!!\n",
    "#Function def used to define a custom function\n",
    "\n",
    "#ex. a greeting function with a changing name\n",
    "def greet (name):\n",
    "    print(\"Hello\", name)\n",
    "\n",
    "greet(\"Alex\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "77ed599d-6843-417d-b46f-17e6e78e863f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n"
     ]
    }
   ],
   "source": [
    "#ex. defining mathematical equation\n",
    "def add(a, b):\n",
    "    return a + b\n",
    "result = add (5, 7)\n",
    "print (result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c30dd4a0-7f98-4273-9f0b-ebebaa7d1e87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n"
     ]
    }
   ],
   "source": [
    "def add(a, b):\n",
    "    return a + b\n",
    "    \n",
    "print (add(5, 7))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cdf4987-73fc-4594-bc9c-7f7607d01ee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Common defined functions in gene analysis\n",
    "__init__ ()\n",
    "calculate_gc_content ()\n",
    "transcribe ()\n",
    "get_coding_region ()\n",
    "__str__ ()\n",
    "__len__ ()\n",
    "\n",
    "#Common defined functions for protein analysis\n",
    "__init__ ()\n",
    "calculate_molecular_weight ()\n",
    "find_amino_acid_position ()\n",
    "has_motif ()\n",
    "__str__ ()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db6316d3",
   "metadata": {},
   "source": [
    "# Importing Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "0f749c68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p-value: 0.00045, -log10(p): 3.35\n",
      "Random DNA sequence: GAAGCTTGTGCAGAGCCTGC\n",
      "Pi value: 3.141592653589793\n",
      "log10(1000) = 3.0\n",
      "e^2 = 7.39\n"
     ]
    }
   ],
   "source": [
    "# Basic imports\n",
    "import math  # Built-in module for mathematical functions\n",
    "import random  # Built-in module for random operations\n",
    "\n",
    "# Calculate probability in bioinformatics context\n",
    "p_value = 0.00045\n",
    "log_p_value = -math.log10(p_value)\n",
    "print(f\"p-value: {p_value}, -log10(p): {log_p_value:.2f}\")\n",
    "\n",
    "# Random DNA sequence generation\n",
    "def generate_random_dna(length):\n",
    "    bases = ['A', 'T', 'G', 'C']\n",
    "    return ''.join(random.choice(bases) for _ in range(length))\n",
    "\n",
    "random_seq = generate_random_dna(20)\n",
    "print(f\"Random DNA sequence: {random_seq}\")\n",
    "\n",
    "# Import with alias (common convention)\n",
    "import math as m\n",
    "print(f\"Pi value: {m.pi}\")\n",
    "\n",
    "# Import specific functions\n",
    "from math import log10, sqrt\n",
    "print(f\"log10(1000) = {log10(1000)}\")\n",
    "\n",
    "# Import everything (not recommended for large modules)\n",
    "from math import *\n",
    "print(f\"e^2 = {exp(2):.2f}\")\n",
    "\n",
    "# Create our own module and import it\n",
    "# First, let's create a simple bioinformatics module in a separate cell\n",
    "# In practice, this would be in a separate file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "e0ccbdcd-5d84-4b79-be38-feecb7cc4261",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.0\n",
      "p-value: 0.00045, -log10(p): 3.35\n",
      "Pi value: 3.141592653589793\n"
     ]
    }
   ],
   "source": [
    "#!!!My code!!!\n",
    "#Module = file which contains Python code / specific functions / packages\n",
    "\n",
    "# import module_name\n",
    "\n",
    "#import math\n",
    "import math\n",
    "print(math.sqrt(16))\n",
    "\n",
    "p_value = 0.00045\n",
    "log_p_value = -math.log10(p_value)\n",
    "print(f\"p-value: {p_value}, -log10(p): {log_p_value:.2f}\")\n",
    "\n",
    "#can name imported functions as you want\n",
    "import math as m\n",
    "print(f\"Pi value: {m.pi}\")\n",
    "\n",
    "#import random\n",
    "import random\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ddff063-3082-459f-b0b2-846dc5cbcffd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de992355-c89e-46a5-93c3-59204f884a20",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
